### PR TITLE
lib/model: Do file recheck in folder loop (fixes #6583)

### LINKED ISSUE #6583: Distributed deadlock on request
This is on a new level

```
goroutine 11193434 [select, 63 minutes]:
github.com/syncthing/syncthing/lib/protocol.(*rawConnection).Request(0xc0113e35f0, 0x10e13c0, 0xc001495880, 0xc0001815e0, 0xb, 0xc010c99340, 0x19, 0x0, 0xd, 0xc010c99380, ...)
	/opt/tcagent/syncthing-4-work/174e136266f8a219/lib/protocol/protocol.go:312 +0x317
github.com/syncthing/syncthing/lib/protocol.wireFormatConnection.Request(0x10ec580, 0xc0113e35f0, 0x10e13c0, 0xc001495880, 0xc0001815e0, 0xb, 0xc010c99340, 0x19, 0x0, 0xd, ...)
	/opt/tcagent/syncthing-4-work/174e136266f8a219/lib/protocol/wireformat.go:40 +0x110
github.com/syncthing/syncthing/lib/model.(*model).requestGlobal(0xc004449500, 0x10e13c0, 0xc001495880, 0x15ccd157e814fe7a, 0xf7a3e7d41cef92de, 0x7df45d739700fead, 0x2847700d8b97c0b2, 0xc0001815e0, 0xb, 0xc010c99340, ...)
	/opt/tcagent/syncthing-4-work/174e136266f8a219/lib/model/model.go:2054 +0x492
github.com/syncthing/syncthing/lib/model.(*sendReceiveFolder).pullBlock(0xc000516900, 0xc001caf900, 0xc010c99380, 0x20, 0x20, 0x0, 0x12a302a80000000d, 0xc01013e3c0)
	/opt/tcagent/syncthing-4-work/174e136266f8a219/lib/model/folder_sendrecv.go:1458 +0x3a8
github.com/syncthing/syncthing/lib/model.(*sendReceiveFolder).pullerRoutine.func1(0x10ded80, 0xc0100c0e70, 0xc00e771e60, 0xd, 0xc000516900, 0xc001caf900, 0xc010c99380, 0x20, 0x20, 0x0, ...)
	/opt/tcagent/syncthing-4-work/174e136266f8a219/lib/model/folder_sendrecv.go:1405 +0xfb
created by github.com/syncthing/syncthing/lib/model.(*sendReceiveFolder).pullerRoutine
	/opt/tcagent/syncthing-4-work/174e136266f8a219/lib/model/folder_sendrecv.go:1401 +0x32d
```

```
goroutine 11152826 [select, 63 minutes]:
github.com/syncthing/syncthing/lib/model.(*folder).Scan(0xc000516900, 0xc00d88c480, 0x1, 0x1, 0xffffffffffffffff, 0xc0019b9208)
	/opt/tcagent/syncthing-4-work/174e136266f8a219/lib/model/folder.go:232 +0x149
github.com/syncthing/syncthing/lib/model.(*folder).ForceRescan(0xc000516900, 0xc00d8d0300, 0x19, 0x0, 0x5ea72c0d, 0x37ee1cad0211b944, 0xc00d8d0320, 0x2, 0x2, 0x299b36, ...)
	/opt/tcagent/syncthing-4-work/174e136266f8a219/lib/model/folder.go:817 +0x16f
github.com/syncthing/syncthing/lib/model.(*model).recheckFile(0xc004449500, 0x15ccd157e814fe7a, 0xf7a3e7d41cef92de, 0x7df45d739700fead, 0x2847700d8b97c0b2, 0x10f1980, 0xc005bc50d0, 0xc00ee76000, 0xb, 0xc00ee685a0, ...)
	/opt/tcagent/syncthing-4-work/174e136266f8a219/lib/model/model.go:1640 +0x2db
github.com/syncthing/syncthing/lib/model.(*model).Request(0xc004449500, 0x15ccd157e814fe7a, 0xf7a3e7d41cef92de, 0x7df45d739700fead, 0x2847700d8b97c0b2, 0xc00ee76000, 0xb, 0xc00ee685a0, 0x19, 0xc00000000d, ...)
	/opt/tcagent/syncthing-4-work/174e136266f8a219/lib/model/model.go:1569 +0x1a71
github.com/syncthing/syncthing/lib/protocol.(*rawConnection).handleRequest(0xc0113e35f0, 0x127, 0xc00ee76000, 0xb, 0xc00ee685a0, 0x19, 0x0, 0xd, 0xc00ee685c0, 0x20, ...)
	/opt/tcagent/syncthing-4-work/174e136266f8a219/lib/protocol/protocol.go:627 +0x112
created by github.com/syncthing/syncthing/lib/protocol.(*rawConnection).dispatcherLoop
	/opt/tcagent/syncthing-4-work/174e136266f8a219/lib/protocol/protocol.go:430 +0x656
```

1. The Scan() issued by recheck during a request can hang until the pull is finished if request comes in during a pull
2. A pull can hang as long as the other side has not responded to a request
3. The other side can not respond to a request because it's trying to do a recheck which is blocked by a scan.

Goto 1.

Seems that all you need is device A to pull some file which hits the recheck condition on device B
and device B to pull some file that hits the recheck condition on device A at the same time.

### PR DESCRIPTION
**Pull request by AudriusButkevicius**
See original ticket for explanation

### CODE REVIEW

**Comment by calmh** [comment]
We're going to get a bunch of retries, ending up in a bunch of queued recheck requests, each of will be handled in order because it's done synchronously in the folder runner. Except for the invalidations in the database which will happen directly... So maybe only the first scan will result in an actual rehash? It's not optimal... Not obvious what the better fix is without going all in on the complexity.

**Comment by AudriusButkevicius** [comment]
I was also wondering about the race in doing the db update, as we could blat over whatever a scan has produced, by inserting a value we've read before the scan inserted it's own value.

**Comment by AudriusButkevicius** [comment]
I think the idea of things modifying local files outside of the scan loop is probably a no-no?
I know we have forceRescan stuff in the puller as well, which potentially suffers from a similar issue?

**Comment by calmh** [comment]
Maybe some of the checks in recheckFile should be pushed into the forcescan method in the folder runner. It can verify that we’re still talking about the version of the file that is in database and if so invalidate it and rescan. That’s just going to happen once then, no matter home many times we ask for it

**Comment by AudriusButkevicius** [comment]
The rescan bit still requires setting, which is still a db operation, and its still racing with the scanner.

**Comment by imsodin** [comment]
yeah this probably needs some tying into the folder loop. we could just store a map of to be invalidated/rehashed files and process that in the loop and remove entries if they change on (normal) scans

**Comment by AudriusButkevicius** [comment]
This now wires it into the folder loop.

File: lib/model/folder.go [hunk 1]
```diff
@@ -921,6 +935,40 @@ func (f *folder) emitDiskChangeEvents(fs []protocol.FileInfo, typeOfEvent events
 	}
 }
 
+func (f *folder) forceRescan() {
+	paths := make([]string, 0, len(f.forcedRescanPaths))
+
+	f.forcedRescanPathsMut.Lock()
+	for path := range f.forcedRescanPaths {
+		paths = append(paths, path)
+		delete(f.forcedRescanPaths, path)
```
Comment (calmh) [inline lib/model/folder.go:944]: I suggest just throwing away the map and creating a new one after the loop, as I don't think map structures shrink once they've grown.

File: lib/model/folder.go [hunk 2]
```diff
@@ -921,6 +935,40 @@ func (f *folder) emitDiskChangeEvents(fs []protocol.FileInfo, typeOfEvent events
 	}
 }
 
+func (f *folder) forceRescan() {
+	paths := make([]string, 0, len(f.forcedRescanPaths))
```
Comment (calmh) [inline lib/model/folder.go:939]: read of forcedRescanPaths must be inside mutex

**Comment by imsodin** [comment]
maybe consider taking this opportunity to rename to forcerehash, seems more accurate.

File: lib/model/folder.go [hunk 3]
```diff
@@ -921,6 +935,39 @@ func (f *folder) emitDiskChangeEvents(fs []protocol.FileInfo, typeOfEvent events
...
+		if !ok {
+			continue
+		}
+		fi.SetMustRescan(f.shortID)
+		batch.append(fi)
+	}
+
+	_ = batch.flush()
+
+	_ = f.scanSubdirs(paths)
```
[the remark this answers is not in the document]
Comment (AudriusButkevicius) [reply lib/model/folder.go:968]: You mean the snap?

[the remark this answers is not in the document]
Comment (imsodin) [reply lib/model/folder.go:969]: yeah

**Pull request merged**